dask/array/linalg.py

```diff
@@ -447,3 +446,77 @@ def lu(a):
     u = Array(dsk, name_u, shape=a.shape, chunks=a.chunks, dtype=a.dtype)
 
     return p, l, u
+
+
+def solve_triangular(a, b, lower=False):
+    """
+    Solve the equation `a x = b` for `x`, assuming a is a triangular matrix.
+
+    Parameters
+    ----------
+    a : (M, M) array_like
+        A triangular matrix
+    b : (M,) or (M, N) array_like
+        Right-hand side matrix in `a x = b`
+    lower : bool, optional
+        Use only data contained in the lower triangle of `a`.
+        Default is to use upper triangle.
+
+    Returns
+    -------
+    x : (M,) or (M, N) array
+        Solution to the system `a x = b`. Shape of return matches `b`.
+    """
+
+    import scipy.linalg
+
+    if a.ndim != 2:
+        raise ValueError('a must be 2 dimensional')
+    if b.ndim <= 2:
+        if a.shape[1] != b.shape[0]:
+            raise ValueError('a.shape[1] and b.shape[0] must be equal')
+        if a.chunks[1] != b.chunks[0]:
+            msg = ('a.chunks[1] and b.chunks[0] must be equal. '
+                   'Use .rechunk method to change the size of chunks.')
+            raise ValueError(msg)
+    else:
+        raise ValueError('b must be 1 or 2 dimensional')
+
+    vchunks = len(a.chunks[1])
+    hchunks = 1 if b.ndim == 1 else len(b.chunks[1])
+    token = tokenize(a, b, lower)
+    name = 'solve-triangular-' + token
+
+    def _b_init(i, j):
+        if b.ndim == 1:
+            return b.name, i
+        else:
+            return b.name, i, j
+
+    def _key(i, j):
+        if b.ndim == 1:
+            return name, i
+        else:
+            return name, i, j
+
+    dsk = {}
+    if lower:
+        for i in range(vchunks):
+            for j in range(hchunks):
+                target = _b_init(i, j)
+                if i > 0:
+                    prevs = (sum, [(np.dot, (a.name, i, k), _key(k, j)) for k in range(i)])
```

Comment:
We probably want to break this up into different tasks for each matrix multiply.

Whenever there is a nested task like `(add, (add, a, b), (add, c, d))` all of `a, b, c, d` will be brought into memory at once to perform the computation.  

For our solve computation this means that all of `(a.name, i, k) k in range(i)` will be brought into memory at once.  In the Matrix-Vector multiply case I expect the result of `(np.dot, (a.name, i, k), _key(k, j))` to be much smaller than the size of the inputs, `(a.name, i, k), _key(k, j)` so it might make sense to make each `np.dot` call a separate task.  This probably doesn't help as much in the matrix-matrix multiply case though.